java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -133,30 +135,71 @@ public void scan(Iterable<InputFile> sourceFiles, Iterable<InputFile> testFiles,
     }
   }
 
-  private void scanAsBatch(Iterable<InputFile>... sourceFiles) {
+  /**
+   * Scans the files given as input in batch mode.
+   *
+   * The batch size used is determined by configuration and bounded to max at 10% of total memory.
+   * This batch size is then used as a threshold: files are added to a batch until the threshold is passed.
+   * Once the threshold is passed, the batch is processed for analysis.
+   *
+   * If no batch size is configured, the input files are scanned as a single batch.
+   *
+   * @param inputFiles The collections of files to scan
+   */
+  private void scanAsBatch(Iterable<? extends InputFile>... inputFiles) {
+    List<InputFile> files = new ArrayList<>();
+    for (Iterable<? extends InputFile> group: inputFiles) {
+      files.addAll(astScanner.filterModuleInfo(group));
+    }
+
+    // Compute the batch size
+    long minBatchModeSize = MAX_BATCH_SIZE;
+    long batchModeSizeInMB = sonarComponents.getBatchModeSizeInMB();
+    if (batchModeSizeInMB != -1L) {
+      minBatchModeSize = batchModeSizeInMB * 1_000_000L;
+      // Bound the memory usage per batch to a maximum of 10% of total memory
+      minBatchModeSize = Math.min(minBatchModeSize, MAX_BATCH_SIZE);
+    }
+    LOG.debug("Scanning with batch size {} B", minBatchModeSize);
+
     try {
-      List<InputFile> allFiles = new ArrayList<>();
-      Arrays.stream(sourceFiles).forEach(files -> files.forEach(allFiles::add));
-      try {
-        JParserConfig.Mode.BATCH
-          .create(JParserConfig.effectiveJavaVersion(javaVersion), globalClasspath)
-          .parse(allFiles, this::analysisCancelled, this::scanAsBatchCallback);
-      } finally {
-        astScanner.endOfAnalysis();
-        astScannerForTests.endOfAnalysis();
-        astScannerForGeneratedFiles.endOfAnalysis();
+      int batchFirstPos = 0;
+      while (batchFirstPos < files.size()) {
+        long batchSize = files.get(batchFirstPos).contents().length();
+        int batchLastPosExclusive = batchFirstPos + 1;
+        while (batchSize < minBatchModeSize && batchLastPosExclusive < files.size()) {
+          batchSize += files.get(batchLastPosExclusive).contents().length();
```

Comment:
Because we always want all the files in AutoScan mode, we should not load the content of all source files in memory at the same time, in this case, it's useless.
And when we are not in the AutoScan context we should exclude batch mode if `sonarComponents.isSonarLintContext()` so we can compute the size using `new File(inputFile.absolutePath()).length()`.